### PR TITLE
Headline AB testing

### PR DESCRIPTION
**Pull request by dominickendrick**
# Headline AB testing
## What does this change?

This adds the ability to test an alternate headline on the US front
## What is the value of this and can you measure success?

The aim of this work is to understand if we can run short AB tests with headlines quickly enough to get significant data to choose the best performing headline.
Success will be collecting enough data to know if this approach is viable.
## Does this affect other platforms - Amp, Apps, etc?

Yes, A Fastly update is required
https://github.com/guardian/fastly-edge-cache/pull/583

### CODE REVIEW

File: common/app/mvt/MultiVariateTesting.scala [hunk 1]
```diff
@@ -25,14 +25,39 @@ object CMTopBannerPosition extends TestDefinition(
...
+}
+
+object ABHeadlinesTestControl extends TestDefinition(
+  Nil,
+  "headlines-ab-control",
+  "To test how much of a difference changing a headline makes (control group)",
+  new LocalDate(2016, 3, 30)
+  ) {
+  override def isParticipating(implicit request: RequestHeader): Boolean = {
+      request.headers.get("X-GU-hlt").contains("hlt-B") && switch.isSwitchedOn && ServerSideTests.isSwitchedOn
```
Comment (OliverJAsh) [inline common/app/mvt/MultiVariateTesting.scala]: Is there a corresponding PR in another project that adds these?

File: common/app/mvt/MultiVariateTesting.scala [hunk 2]
```diff
@@ -25,14 +25,39 @@ object CMTopBannerPosition extends TestDefinition(
...
   def getJavascriptConfig(implicit request: RequestHeader): String = {
 
-    val configEntries = List(InternationalEditionVariant(request).map{ international => s""""internationalEditionVariant" : "$international" """}) ++
-      List(ActiveTests.getParticipatingTest(request).map{ test => s""""${CamelCase.fromHyphenated(test.name)}" : ${test.switch.isSwitchedOn}"""})
+    val headlineTests = List(ABHeadlinesTestControl, ABHeadlinesTestVariant).filter(_.isParticipating)
+          .map{ test => Some(s""""${CamelCase.fromHyphenated(test.name)}" : ${test.switch.isSwitchedOn}""")}
 
+    val configEntries = List(InternationalEditionVariant(request).map{ international => s""""internationalEditionVariant" : "$international" """}) ++
+      List(ActiveTests.getParticipatingTest(request).map{ test => s""""${CamelCase.fromHyphenated(test.name)}" : ${test.switch.isSwitchedOn}"""}) ++
+      headlineTests
```
Comment (OliverJAsh) [inline common/app/mvt/MultiVariateTesting.scala]: Minor note but you could remove the duplication here:

``` scala
val headlineTests = List(ABHeadlinesTestControl, ABHeadlinesTestVariant).filter(_.isParticipating)

val configEntries = List(InternationalEditionVariant(request).map{ international => s""""internationalEditionVariant" : "$international" """}) ++
  (headlineTests ++ List(ActiveTests.getParticipatingTest(request)).map{ test => s""""${CamelCase.fromHyphenated(test.name)}" : ${test.switch.isSwitchedOn}"""})
```

**Comment by OliverJAsh** [comment]
Looks great! :+1:

File: common/app/headlines/ABHeadlines.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  def refresh(): Unit = if (ABHeadlinesTestVariant.switch.isSwitchedOn) {
+    Future {
+      blocking {
+        val spreadsheet = service.getFeed(spreadsheetUrl, classOf[WorksheetFeed])
+        spreadsheet.getEntries.headOption.foreach { worksheet =>
+          val feed = service.getFeed(worksheet.getCellFeedUrl, classOf[CellFeed])
+          val cells = feed.getEntries
+          if (cells.length == 5) {  //if it is a different value someone has messed with the spreadsheet
+          val testHeadline = cells(3).getCell.getValue -> cells(4).getCell.getValue
```
Comment (johnduffell) [inline common/app/headlines/ABHeadlines.scala]: watch the indentation here

File: common/app/headlines/ABHeadlines.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  private def isUsFront(req: RequestHeader) = req.path == "/us"
+}
+
+trait ABHeadlinesLifecycle extends GlobalSettings {
+
+  private val ABHeadlinesJob = "ABHeadlinesJob"
+
+  override def onStart(app: play.api.Application) {
+    super.onStart(app)
+    Jobs.schedule(ABHeadlinesJob, "0 * * * * ?") {
```
Comment (johnduffell) [inline common/app/headlines/ABHeadlines.scala]: is it possible to use a "scheduleEveryNMinutes" call or something?  this string isn't self explanatory so very difficult to review

File: common/app/mvt/MultiVariateTesting.scala [hunk 5]
```diff
@@ -25,14 +25,39 @@ object CMTopBannerPosition extends TestDefinition(
...
 )
 
+object ABHeadlinesTestVariant extends TestDefinition(
+  Nil,
+  "headlines-ab-variant",
+  "To test how much of a difference changing a headline makes (variant group)",
+  new LocalDate(2016, 3, 30)
+  ) {
+  override def isParticipating(implicit request: RequestHeader): Boolean = {
+    request.headers.get("X-GU-hlt").contains("hlt-A") && switch.isSwitchedOn && ServerSideTests.isSwitchedOn
```
Comment (johnduffell) [inline common/app/mvt/MultiVariateTesting.scala]: why not hlt-v and hlt-c for variant and control to make it consistent?

File: common/app/mvt/MultiVariateTesting.scala (hunk 1 above)
Comment (johnduffell) [inline common/app/mvt/MultiVariateTesting.scala]: yes these will be added at the CDN

File: static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    function clearHash() {
+        if (detect.hasPushStateSupport()) {
+            history.replaceState('', document.title, window.location.pathname);
+        } else {
+            window.location.hash = '';
+        }
+    }
+
+    return {
+        go: function () {
```
Comment (jamesgorrie) [inline static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js]: `init` is the consistent way as far as I can tell?

File: static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            var hash = window.location.hash;
+
+            if (window.location.pathname === '/us') {
```
Comment (jamesgorrie) [inline static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js]: You could do something like `page.contentType === 'Network Front' && page.edition === 'us'`.
I don't mind though,.

**Comment by johnduffell** [comment]
apart from comments, exciting stuff!! :+1:

File: static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js (hunk 6 above)
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/analytics/headlines-test-analytics.js]: Why not just return the single function?

**Comment by jbreckmckye** [comment]
This is really exciting! :8ball:

**Comment by gklopper** [comment]
Looking forward to this kicking off again

**Comment by gklopper** [comment]
> The aim of this work is to understand if we can run short AB tests with headlines quickly enough to get significant data to choose the best performing headline.

Just to add to this, the real prize is if we can get a better headline/ link text, then that lives with content forever more, increasing the chance someone will read it every time it shows up somewhere.

**Pull request merged**